blocks/blockoracle.go

```diff
@@ -60,23 +60,23 @@ func NewMinerBlockOracle(committeeSize uint32, genesisActiveSetSize uint32, laye
 
 // BlockEligible returns the ATXID and list of block eligibility proofs for the given layer. It caches proofs for a
 // single epoch and only refreshes the cache if eligibility is queried for a different epoch.
-func (bo *Oracle) BlockEligible(layerID types.LayerID) (types.ATXID, []types.BlockEligibilityProof, error) {
+func (bo *Oracle) BlockEligible(layerID types.LayerID) (types.ATXID, []types.BlockEligibilityProof, []types.ATXID, error) {
 	if !bo.isSynced() {
-		return types.ATXID{}, nil, fmt.Errorf("cannot calc eligibility, not synced yet")
+		return types.ATXID{}, nil, nil, fmt.Errorf("cannot calc eligibility, not synced yet")
 	}
 	epochNumber := layerID.GetEpoch()
 	bo.log.Info("asked for eligibility for epoch %d (cached: %d)", epochNumber, bo.proofsEpoch)
 	if epochNumber.IsGenesis() {
 		bo.log.Warning("asked for eligibility for epoch 0, cannot create blocks here")
-		return *types.EmptyATXID, nil, nil
+		return *types.EmptyATXID, nil, nil, nil
 	}
-	if bo.proofsEpoch != epochNumber {
-		err := bo.calcEligibilityProofs(epochNumber)
-		if err != nil {
-			bo.log.Error("failed to calculate eligibility proofs for epoch %v : %v", epochNumber, err)
-			return *types.EmptyATXID, nil, err
-		}
+
+	activeSet, err := bo.calcEligibilityProofs(epochNumber)
+	if err != nil {
+		bo.log.Error("failed to calculate eligibility proofs for epoch %v : %v", epochNumber, err)
+		return *types.EmptyATXID, nil, nil, err
```

Comment:
This may be problematic. It essentially cancels caching and calculates the miner's eligibility every time this method is called. Calculating the eligibility is very expensive. It takes several seconds with testnet parameters (288 layers per epoch with 50 blocks each = 14,400 blocks per epoch) it will be even heavier in mainnet (14x = 201,600 blocks per epoch).

The plan that I had in mind was to calculate this during the last layer of the epoch for the next epoch. Doing the calculation in real-time at the start of each layer is wasteful and not practical.

If you want to know the `activeSet` you should cache it during eligibility proof calculation and get it from the cache instead of canceling the cache.

Reply:
the active set is not calculated using mesh traversal so calculating active set size becomes a matter of reading ATX ids from database

Reply:
`calcEligibilityProofs` isn't expensive because of traversal, it performs a lot of signing. The new algo in var post may be more efficient, but I haven't benchmarked it yet, so can't say for sure.

I think that adding the active set to the cache is trivial - let's do that.

Reply:
fixed, pls approve :)